rust/dns-over-tcp/src/server.rs

```diff
@@ -0,0 +1,281 @@
+        );
+
+        if !changed {
+            return;
+        }
+
+        for (handle, smoltcp::socket::Socket::Tcp(socket)) in self.sockets.iter_mut() {
+            let listen = self.listen_endpoints.get(&handle).copied().unwrap();
+
+            match try_recv_message(socket, listen) {
+                Ok(Some(message)) => {
+                    self.received_queries.push_back(Query {
+                        message,
+                        socket: SocketHandle(handle),
+                        local: listen,
+                    });
+                }
+                Ok(None) => {}
+                Err(e) => {
+                    tracing::debug!("Failed recv message: {e}");
+                    socket.abort();
+                }
+            }
+        }
+    }
+
+    pub fn poll_outbound(&mut self) -> Option<IpPacket> {
+        self.device.outbound_packets.pop_front()
+    }
+
+    pub fn poll_queries(&mut self) -> Option<Query> {
+        self.received_queries.pop_front()
+    }
+}
+
+fn create_tcp_socket(listen_endpoint: SocketAddr) -> tcp::Socket<'static> {
+    /// The 2-byte length prefix of DNS over TCP messages limits their size to effectively u16::MAX.
+    /// It is quite unlikely that we have to buffer _multiple_ of these max-sized messages.
+    /// Being able to buffer at least one of them means we can handle the extreme case.
+    /// In practice, this allows the OS to queue multiple queries even if we can't immediately process them.
+    const MAX_TCP_DNS_MSG_LENGTH: usize = u16::MAX as usize;
+
+    let mut socket = tcp::Socket::new(
+        RingBuffer::new(vec![0u8; MAX_TCP_DNS_MSG_LENGTH]),
+        RingBuffer::new(vec![0u8; MAX_TCP_DNS_MSG_LENGTH]),
+    );
+    socket
+        .listen(listen_endpoint)
+        .expect("A fresh socket should always be able to listen");
+
+    socket
+}
+
+fn try_recv_message(
+    socket: &mut tcp::Socket,
+    listen: SocketAddr,
+) -> Result<Option<Message<Vec<u8>>>> {
+    // smoltcp's sockets can only ever handle a single remote, i.e. there is no permanent listening socket.
+    // to be able to handle a new connection, reset the socket back to `listen` once the connection is closed / closing.
+    {
+        use smoltcp::socket::tcp::State::*;
+
+        if matches!(socket.state(), Closed | TimeWait | CloseWait) {
+            tracing::debug!("Resetting socket to listen state");
+
+            socket.abort();
+            socket
+                .listen(listen)
+                .expect("Can always listen after `abort()`");
+        }
+    }
+
+    // We configure `smoltcp` with "any-ip", meaning packets to technically any IP will be routed here to us.
+    if let Some(local) = socket.local_endpoint() {
+        if local != IpEndpoint::from(listen) {
+            bail!("Bad destination socket: {local}");
+        }
+    }
+
+    // Ensure we can recv, send and have space to send.
+    if !socket.can_recv() || !socket.can_send() || socket.send_queue() > 0 {
```

Comment:
Do we need this check? I think `recv` fails if socket can't recieve, and I'm not sure why we need the others

Reply:
These checks basically ensure that we only read if we can also send. If an application would spam us with DNS queries over a single socket, this makes sure that it will only fill up the buffer and backpressure to the application instead of us accepting more queries before answering existing ones.